### ISSUE TITLE
Order of evaluation during construction

### ISSUE THREAD

**Issue by dddejan** [SRI-CSL/solidity#24]
Running on the file now I get:

```
dejan$ ./solc-verify.py test/solc-verify/Issue003.sol 
Verifying A::[implicit_constructor]: OK
Verifying Issue003::[constructor]: ERROR
 - Source test/solc-verify/Issue003.sol, line 149, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 150, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 151, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 152, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 153, col 9: Assertion might not hold.
Verifying Issue003::: ERROR
 - Source test/solc-verify/Issue003.sol, line 149, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 150, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 151, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 152, col 9: Assertion might not hold.
 - Source test/solc-verify/Issue003.sol, line 153, col 9: Assertion might not hold.
Errors were found by the verifier.
```

This is unexpected.

_Originally posted by @dddejan in https://github.com/SRI-CSL/solidity/issues/3#issuecomment-499630181_

**Comment by hajduakos** [comment]
Interestingly, this only holds for constants:
```
contract Test {
    int constant x = y;
    int constant y = 5;

    function() external payable {
        assert(x == 5);
        assert(y == 5);
    }
}
```

For non-constant variables the order seems to be the same as they are defined (e.g., `x` is initialized with the yet uninitialized value `0` of `y`).

```
contract Test {
    int x = y;
    int y = 5;

    function() external payable {
        assert(x == 0);
        assert(y == 5);
    }
}
```

The compiler can warn about cyclical dependencies between constants, so there should be some way of getting this info and using it to initialize constants.

**Comment by dddejan** [comment]
Yeah, seems so. Fairly new for now.

https://solidity.readthedocs.io/en/v0.5.9/contracts.html#constant-state-variables

**Comment by hajduakos** [comment]
There is a `PostTypeChecker` class which internally builds a dependency graph for constants and checks for cycles. However, this information does not seem to be persisted for later passes (e.g., as AST annotations). Based on the documentation, every occurrence of the constant is simply replaced by its respective initializer expression. We can also do the same in the translation: we do not introduce Boogie variables for the constants, but rather put them and their expressions in some mapping. Then, if we see an identifier referencing a constant, we look the expression up in the map and translate it to Boogie.

**Comment by hajduakos** [comment]
We don't actually even need a map. If an identifier references a constant, we can get the declaration and get its initializer expression. Then, instead of translating the identifier, we translate the expression.

**Comment by hajduakos** [comment]
Implemented inlining in c44b9091851557c3984065e86b447a0b0ab1da84

**Comment by dddejan** [comment]
Great, the inlining is exactly what's needed and makes the checks pass even for the main function!
